pkg/controlplane/controller_test.go

```diff
@@ -17,430 +17,17 @@ limitations under the License.
-		if err != nil {
-			t.Errorf("case %q: unexpected error: %v", test.testName, err)
-		}
-		updates := []core.UpdateAction{}
-		for _, action := range fakeClient.Actions() {
-			if action.GetVerb() == "update" {
-				updates = append(updates, action.(core.UpdateAction))
-			}
-		}
-		if test.expectUpdate != nil {
-			if len(updates) != 1 {
-				t.Errorf("case %q: unexpected updates: %v", test.testName, updates)
-			} else {
-				obj := updates[0].GetObject()
-				if e, a := test.expectUpdate.Spec, obj.(*corev1.Service).Spec; !reflect.DeepEqual(e, a) {
-					t.Errorf("case %q: expected update:\n%#v\ngot:\n%#v\n", test.testName, e, a)
-				}
-			}
-		}
-		if test.expectUpdate == nil && len(updates) > 0 {
-			t.Errorf("case %q: no update expected, yet saw: %v", test.testName, updates)
-		}
-	}
-
-	nonReconcileTests := []struct {
-		testName     string
-		serviceName  string
-		servicePorts []corev1.ServicePort
-		serviceType  corev1.ServiceType
-		service      *corev1.Service
-		expectUpdate *corev1.Service // nil means none expected
-	}{
-		{
-			testName:    "service definition wrong port, no expected update",
-			serviceName: "foo",
-			servicePorts: []corev1.ServicePort{
-				{Name: "foo", Port: 8080, Protocol: "TCP", TargetPort: intstr.FromInt32(8080)},
-			},
-			serviceType: corev1.ServiceTypeClusterIP,
-			service: &corev1.Service{
-				ObjectMeta: om("foo"),
-				Spec: corev1.ServiceSpec{
-					Ports: []corev1.ServicePort{
-						{Name: "foo", Port: 1000, Protocol: "TCP", TargetPort: intstr.FromInt32(1000)},
-					},
-					Selector:        nil,
-					ClusterIP:       "1.2.3.4",
-					SessionAffinity: corev1.ServiceAffinityNone,
-					Type:            corev1.ServiceTypeClusterIP,
-				},
-			},
-			expectUpdate: nil,
-		},
-	}
-	for _, test := range nonReconcileTests {
-		master := Controller{}
-		fakeClient := fake.NewSimpleClientset(test.service)
-		master.client = fakeClient
-		err := master.CreateOrUpdateMasterServiceIfNeeded(test.serviceName, netutils.ParseIPSloppy("1.2.3.4"), test.servicePorts, test.serviceType, false)
-		if err != nil {
-			t.Errorf("case %q: unexpected error: %v", test.testName, err)
-		}
-		updates := []core.UpdateAction{}
-		for _, action := range fakeClient.Actions() {
-			if action.GetVerb() == "update" {
-				updates = append(updates, action.(core.UpdateAction))
-			}
-		}
-		if test.expectUpdate != nil {
-			if len(updates) != 1 {
-				t.Errorf("case %q: unexpected updates: %v", test.testName, updates)
-			} else {
-				obj := updates[0].GetObject()
-				if e, a := test.expectUpdate.Spec, obj.(*corev1.Service).Spec; !reflect.DeepEqual(e, a) {
-					t.Errorf("case %q: expected update:\n%#v\ngot:\n%#v\n", test.testName, e, a)
-				}
-			}
-		}
-		if test.expectUpdate == nil && len(updates) > 0 {
-			t.Errorf("case %q: no update expected, yet saw: %v", test.testName, updates)
-		}
-	}
-}
-
 func Test_completedConfig_NewBootstrapController(t *testing.T) {
```

Comment:
this should go into kubernetesservice too

Reply:
Not yet. it depends on a check here. But last commit in https://github.com/kubernetes/kubernetes/pull/119042 will move it where it belongs, namely into kube-apiserver options validation.